automates/model_analysis/graph_manipulation.py

```diff
@@ -0,0 +1,634 @@
+        verts_0 = cg.vs.select(orig_name=old_vert_name0)
+        for vert in verts_0:
+            if vert["val_assign"] is None:
+                unobs_edges_to_add.append((new_vert_indx, vert.index))
+        verts_1 = cg.vs.select(orig_name=old_vert_name1)
+        for vert in verts_1:
+            if vert["val_assign"] is None:
+                unobs_edges_to_add.append((new_vert_indx, vert.index))
+    for node in initial_verts:
+        # if "U" not in parents_unsort(cg_node_info[i].orig_name, cg):
+        i = node.index
+        if not any(i in edge for edge in unobs_edges_to_add):
+            cg.add_vertices(1, attributes={"name": f"U_{node['name']}"})
+            new_vert_indx = cg.vs.select(name=f"U_{node['name']}").indices[0]
+            # Find all nodes across parallel worlds
+            verts_to_connect = cg.vs.select(orig_name=node["name"])
+            for vert in verts_to_connect:
+                if vert["val_assign"] is None:
+                    unobs_edges_to_add.append((new_vert_indx, vert.index))
+    cg.add_edges(unobs_edges_to_add, attributes={"description": ["U"] * len(unobs_edges_to_add)})
+    return cg
+
+
+@dataclass(unsafe_hash=True)
+class Probability:
+    var: list = field(default_factory=list)
+    cond: list = field(default_factory=list)
+    sumset: list = field(default_factory=list)
+    do: str = ""
+    product: bool = False
+    children: list = field(default_factory=list)
+    fraction: bool = False
+    domain: int = 0
+    sum: bool = False
+    weight: list = field(default_factory=list)
+    num: Probability = None
+    den: Probability = None
+
+
+@dataclass(unsafe_hash=True)
+class Call:
+    y: str = ""
+    x: str = ""
+    z: str = ""
+    z_prime: str = ""
+    p: Probability = Probability()
+    g: igraph.Graph = igraph.Graph()
+    line: int = 0
+    v: list = field(default_factory=list)
+    id_check: bool = False
+    ancestors: list = field(default_factory=list)
+    w: list = field(default_factory=list)
+    an_xbar: list = field(default_factory=list)
+    s: list = field(default_factory=list)
+    s_prime: list = field(default_factory=list)
+
+
+@dataclass(unsafe_hash=True)
+class TreeNode:
+    root: Probability = Probability()
+    call: Call = Call()
+    children: List[TreeNode] = field(default_factory=list)
+
+
+@dataclass(unsafe_hash=True)
+class ResultsInternal:
+    p: Probability = Probability()
+    tree: TreeNode = TreeNode()
+
+
+@dataclass(unsafe_hash=True)
+class Results:
+    query: dict = field(default_factory=dict)
+    algorithm: str = ""
+    p: str = ""
+    tree: TreeNode = TreeNode()
+
+
+@dataclass
+class CF:
+    node: str = None
+    val_assign: str = None
+    int_var: str = None
+
+
+@dataclass
+class CGNode:
+    orig_name: str = None
+    val_assign: str = None
+    int_var: str = None
+
+
+class IDANotIdentifiable(Exception):
+    pass
+
+
+gamma = [CF("Y", "y", "X"), CF("X", "x_prime"), CF("Z", "z", "D"), CF("D", "d")]
```

Comment:
@aswinchester please either move this code to a script that you can use for debugging/smoke-testing or move it to your testing file.

Reply:
I have moved this into a script :)